### PR TITLE
Reverse kanji search

### PR DESCRIPTION
**Pull request by Alkhwarizm**
A PR to address issue #46.

It basically modified the kanji search command so that it also accepts query that contains no kanji in it. It scrapes Jisho for kanji related to the query if the query contains no kanji.

What do you think?

### CODE REVIEW

**Comment by mistval** [comment]
Very nice, sounds good to me, but this will send a request to Jisho's kanji result page any time anyone uses the the Jisho command, even if they never click the reaction to see the kanji page (which is most of the time). I'd rather only make that request when the kanji page is accessed. You could do that by moving the code into KanjiNavigationDataSource's `prepareData` function. That only gets called the first time the kanji page is accessed, so you can make the request there and store the kanji in the `this.kanji` array.

The "I didn't find any results thing" would need to move into `getPageFromPreparedData` then. If there are no kanji, you can return that message when `pageIndex` is 0, and undefined otherwise.

**Comment by Alkhwarizm** [comment]
Oh, that makes sense. Now I get how the `prepareData` and `getPageFromPreparedData` works. I will try to do that.

File: bot/src/discord/create_kanji_search_navigation_chapter.js [hunk 1]
```diff
@@ -42,44 +57,24 @@ class KanjiNavigationDataSource {
...
-    kanji,
+    searchQuery,
     authorName,
     commandPrefix,
     forceNavigationFooter,
   );
 
   return {
     navigationChapter: new NavigationChapter(dataSource),
-    pageCount: kanji.length,
```
Comment (Alkhwarizm) [inline bot/src/discord/create_kanji_search_navigation_chapter.js]: I have done as you suggested. But now I'm not quite sure how to get the `pageCount` and `hasResult` that is returned by `createKanjiSearchNavigationChapter`, without accessing the kanji page.  It seems that the real count of the page doesn't really matters. What seems to matter is only whether it is one or more than one page.

**Comment by mistval** [comment]
There was a little more to it than I was thinking, I had to do it to figure out how to do it, since I haven't touched this code in a bit, so I finished it off and committed to your fork, hope you don't mind.

Before this change, we knew how many kanji result pages we had without having to hit jisho.org, since we could just count the kanji in the query. So when you used the k!j command, if there were no kanji results, it didn't add the K reaction button. Now, it doesn't know if there are kanji results or not, so I made it show the K button anyway. It's maybe a little bit worse user experience, but I think it's fine. 

So with that, the Jisho command no longer cares how many pages the kanji results has. It will add the K reaction anyway, and it will add the arrow reactions anyway (since it might need them for the other chapters).

Now only the kanji command cares how many pages the results has, so I added a function to the kanji data source to get the page count asynchronously. Since the kanji command only shows the kanji chapter, the considerations about needlessly querying Jisho.org don't apply to it, since if you use the command you're going to see the kanji results.

**Pull request merged**